### PR TITLE
feat: [websocket] 看見其他玩家加入/離開(#154)"

### PR DESCRIPTION
**Pull request by cooper-car**
## Why need this change? / Root cause: 
- web-socket 
  - user joined room event
  - user leaved room event
## Changes made:
- Room related
## Test Scope / Change impact:
- RoomController
## Issue
- #154

### CODE REVIEW

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 1]
```diff
@@ -57,6 +64,15 @@ class JoinRoomUsecase(
         return roomRepository.update(this)
     }
 
+    private fun Room.joinRoomEvent(
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:67]: Room.joinPlayer 時，回傳 **PlayerJoinedRoomEvent**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/LeaveRoomUsecase.kt [hunk 2]
```diff
@@ -21,9 +26,21 @@ class LeaveRoomUsecase(
...
     }
 
+    private fun Room.leaveRoomEvent(
+        playerId: String,
+        nickname: String
+    ) : UserLeavedRoomEvent {
+        val user = UserInfo(playerId, nickname)
+        val data = Data(user, roomId!!.value)
+        return UserLeavedRoomEvent(USER_LEFT, data)
+    }
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/LeaveRoomUsecase.kt:42]: room.leaveRoom，回傳 **PlayerLeavedRoomEvent**

File: domain/src/main/kotlin/tw/waterballsa/gaas/events/UserJoinedRoomEvent.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+data class UserJoinedRoomEvent(
+    val type: EventMessageType,
+    val data: Data
+) : DomainEvent() {
+    data class Data(
+        val user: UserInfo,
+        val roomId: String
+    ) {
+        data class UserInfo(
```
Comment (Wally5077) [inline domain/src/main/kotlin/tw/waterballsa/gaas/events/UserJoinedRoomEvent.kt:13]: 直接叫 Player 就好~

File: domain/src/main/kotlin/tw/waterballsa/gaas/events/UserLeavedRoomEvent.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+data class UserLeavedRoomEvent(
+    val type: EventMessageType,
+    val data: Data
+) : DomainEvent() {
+    data class Data(
+        val user: UserInfo,
+        val roomId: String
+    ) {
+        data class UserInfo(
```
Comment (Wally5077) [inline domain/src/main/kotlin/tw/waterballsa/gaas/events/UserLeavedRoomEvent.kt:13]: ditto Player

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt (hunk 1 above)
Comment (cooper-car) [reply application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:67]: 已修正

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/LeaveRoomUsecase.kt (hunk 2 above)
Comment (cooper-car) [reply application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/LeaveRoomUsecase.kt]: 已修正

File: domain/src/main/kotlin/tw/waterballsa/gaas/events/UserJoinedRoomEvent.kt (hunk 3 above)
Comment (cooper-car) [reply domain/src/main/kotlin/tw/waterballsa/gaas/events/UserJoinedRoomEvent.kt]: 已修正

File: domain/src/main/kotlin/tw/waterballsa/gaas/events/UserLeavedRoomEvent.kt (hunk 4 above)
Comment (cooper-car) [reply domain/src/main/kotlin/tw/waterballsa/gaas/events/UserLeavedRoomEvent.kt]: 已修正

**Review by frankvicky — APPROVED** [review]

**Pull request merged**